### PR TITLE
[lldb] Correctly annotate threads at a bp site as hitting it

### PR DESCRIPTION
**Pull request by jasonmolenda**
This is next in my series of "fix the racey tests that fail on greendragon" addressing the failure of TestConcurrentManyBreakpoints.py where we set a breakpoint in a function that 100 threads execute, and we check that we hit the breakpoint 100 times.  But sometimes it is only hit 99 times, and the test fails.

When we hit a software breakpoint, the pc value for the thread is the address of the breakpoint instruction - as if it had not been hit yet.  And because a user might ADD a breakpoint for the current pc from the commandline, when we go to resume execution, any thread that is sitting at a breakpoint site will be silently advanced past the breakpoint instruction (disable bp, instruction step that thread, re-enable bp) before resuming -- whether that thread has hit its breakpoint or not.

What this test is exposing is that there is another corner case, a thread that is sitting at a breakpoint site but has not yet executed the breakpoint instruction.  The thread will have no stop reason, no mach exception, so it will not be recorded as having hit the breakpoint (because it hasn't yet).  But when we resume execution, because it is sitting at a breakpoint site, we advance past it and miss the breakpoint hit.

In 2016 Abhishek Aggarwal handled a similar issue with a patch in `ProcessGDBRemote::SetThreadStopInfo()`, adding a breakpoint StopInfo for a thread sitting at a breakpoint site that has no stop reason. debugserver's `jThreadsInfo` would not correctly execute Abhishek's code though because it would respond with `"reason":"none"` for a thread with no stop reason, and `SetThreadStopInfo()` expected an empty reason here.  The first part of my patch is to clear the `reason` if it is `"none"` so we flow through the code correctly.

On Darwin, though, our stop reply packet (Txx...) includes the `threads`, `thread-pcs`, and `jstopinfo` keys, which give us the tids for all current threads, the pc values for those threads, and `jstopinfo` has a JSON dictionary with the mach exceptions for all threads that have a mach exception.  In
`ProcessGDBRemote::CalculateThreadStopInfo()` we set the StopInfo for each thread for a private stop and if we have `jstopinfo` it is the source of all the StopInfos.  I have to add the same logic here, to give the thread a breakpoint StopInfo even though it hasn't executed the breakpoint yet.  In this case we are very early in thread construction and I only have the information in the Txx stop reply packet -- tids, pcs, and jstopinfo, so I can't use the normal general mechanisms of going through the RegisterContext to get the pc, it's a bit different.

If I hack debugserver to not issue `jstopinfo`, `CalculateThreadStopInfo` will fall back to sending `qThreadStopInfo` for each thread and going through `ProcessGDBRemote::SetThreadStopInfo()` to set the stop infos (and with the `reason:none` fix, use Abhishek's code).

rdar://110549165

### CODE REVIEW

**Review by JDevlieghere — COMMENTED** [review]
Makes sense to me, but I'll defer to @jimingham as he's the expert.

File: lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp [hunk 1]
```diff
@@ -1722,6 +1742,8 @@ ThreadSP ProcessGDBRemote::SetThreadStopInfo(
     } else {
       bool handled = false;
       bool did_exec = false;
+      if (reason == "none")
```
Comment (jimingham) [inline lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:1745]: This definitely needs a comment, or maybe somewhere else saying that none == empty, so we always convert none to empty.

**Review by jimingham — APPROVED** [review]
This seems safe to me, and it can't hurt to take care of this corner case here regardless of what the higher levels of lldb does.

It bugs me a little that we're treating what seems like a general problem down in the GDBRemote process plugin.  But anyway, if we are going to make this a process plugin policy: that the plugins are responsible for producing breakpoint stop reasons both for when the underlying system generates one and when we see "artificial" hits like this we should probably say so somewhere.

With the couple doc comments aside, I'm fine with this.

File: lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp (hunk 1 above)
Comment (jasonmolenda) [reply lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp]: I updated the patch where I handle this a little more cleanly, checking that `reason` has a value, and that the value is not `"none"`, and added a short comment.

**Comment by jasonmolenda** [comment]
> This seems safe to me, and it can't hurt to take care of this corner case here regardless of what the higher levels of lldb does.

This all falls under `Thread::CalculateStopInfo()` - `ThreadGDBRemote::CalculateStopInfo` is calling into ProcessGDBRemote where we do this.  The base class method is pure virtual, so I'm not sure there's a higher level place where this logic could be situated.  I updated the PR with a comment on `Thread::CalculateStopInfo()` specifically detailing this issue, so people writing new subclasses can (hopefully) be aware of it.

**Pull request merged**